src/cpp/session/modules/SessionDependencies.cpp

```diff
@@ -291,67 +326,114 @@ Error unsatisfiedDependencies(const json::JsonRpcRequest& request,
    return Success();
 }
 
-Error installDependencies(const json::JsonRpcRequest& request,
-                          json::JsonRpcResponse* pResponse)
+// Builds an installation script which will install all the dependencies at once. 
+std::string buildCombinedInstallScript(const std::vector<Dependency>& deps)
 {
-   // get list of dependencies
-   json::Array depsJson;
-   std::string context;
-   Error error = json::readParams(request.params, &context, &depsJson);
-   if (error)
-      return error;
-   std::vector<Dependency> deps = dependenciesFromJson(depsJson);
+   bool isRenv = module_context::isRenvActive();
+   std::vector<std::string> cranPackages;
+   std::vector<std::string> cranSourcePackages;
+   std::vector<std::string> embeddedPackages;
+   std::string cmd;
+
+   // Sort the dependencies into CRAN packages installed with defaults, CRAN packages explicitly
+   // installed as source, and embedded packages.
+   for (const Dependency& dep: deps)
+   {
+      if (dep.location == kCRANPackageDependency)
+      {
+         if (dep.source)
+            cranSourcePackages.push_back("'" + dep.name + "'");
+         else
+            cranPackages.push_back("'" + dep.name + "'");
+      }
+      else if (dep.location == kEmbeddedPackageDependency)
+      {
+         EmbeddedPackage pkg = embeddedPackageInfo(dep.name);
+         if (!pkg.empty())
+            embeddedPackages.push_back(pkg.archivePath);
+      }
+   }
 
-   // Ensure we have a writeable user library
-   error = r::exec::RFunction(".rs.ensureWriteableUserLibrary").call();
-   if (error)
-      return error;
+   // Install the CRAN packages with a single call
+   if (!cranPackages.empty())
```

Comment:
What happens if a CRAN package depends on a CRAN source package (ie: something installed in the next step)?

Reply:
A good question! That would theoretically be a problem, but this is logic copied wholesale from 1.2 in an attempt to reduce risk, and so my inclination is to leave it undisturbed unless we have evidence it's causing real issues (practically speaking we rarely insist on source packages so I don't think this has happened often, if ever).